### PR TITLE
Implement cell semantics

### LINKED ISSUE #55: Implement Notebook Cell Semantics
Notes:
- [x] New cells are always code cells and are started in edit mode
- [x] Executing a cell selects the next cell but does not focus it
- [x] If the cell is not focused, we are in command mode
- [x] Cells can be executed in command mode
- [x] A cell can be: selected, focused, and marked, each of which has a different semantic meaning
- [x] Rendering an empty markdown cell should display a message:
  Type Markdown and LaTeX:  α^2

### PR DESCRIPTION
**Pull request by blink1073**
jupyter-lab-cells

Fixes #55.

I'll save cell marking for later.

### CODE REVIEW

**Comment by blink1073** [comment]
ping @jasongrout @Carreau @minrk

**Comment by jasongrout** [comment]
Thanks.  Looking at the code now.

File: src/cells/model.ts [hunk 1]
```diff
@@ -276,10 +306,21 @@ class BaseCellModel implements IBaseCellModel {
   type: CellType;
 
   /**
-   * Re-emit changes to the input dirty state.
+   * Handle changes to the input model.
+   */
+  private onInputChanged(input: IInputAreaModel, args: IChangedArgs<any>): void {
```
Comment (jasongrout) [inline src/cells/model.ts]: I think the convention is that private things always start with `_`.

File: src/cells/model.ts [hunk 2]
```diff
@@ -276,10 +306,21 @@ class BaseCellModel implements IBaseCellModel {
   type: CellType;
 
   /**
-   * Re-emit changes to the input dirty state.
+   * Handle changes to the input model.
+   */
+  protected onInputChanged(input: IInputAreaModel, args: IChangedArgs<any>): void {
+    // Re-emit changes to input dirty and readOnly states.
+    if (args.name === 'dirty' || args.name === 'readOnly' || args.name ) {
```
Comment (jasongrout) [inline src/cells/model.ts]: Won't this re-emit _every_ input area change now (because of the `|| args.name`)? Is that intentional?

Comment (blink1073) [inline src/cells/model.ts]: Fixed, thanks.

**Comment by dwillmer** [comment]
:+1:

File: src/cells/model.ts [hunk 3]
```diff
@@ -182,6 +192,26 @@ class BaseCellModel implements IBaseCellModel {
...
+   *
+   * #### Notes
+   * This is a delegate to the focused state of the input's editor.
+   */
+  get focused(): boolean {
+    return this.input.textEditor.focused;
+  }
+
+  /**
+   * Get whether the cell is focused for editing.
```
Comment (dwillmer) [inline src/cells/model.ts]: Should be `set`

File: src/cells/widget.ts [hunk 4]
```diff
@@ -73,6 +78,16 @@ const RAW_CELL_CLASS = 'jp-RawCell';
  */
 const RENDERED_CLASS = 'jp-mod-rendered';
 
+/**
+ * The class name added to a fucused cell.
```
Comment (dwillmer) [inline src/cells/widget.ts]: focused.

File: src/cells/model.ts (hunk 3 above)
Comment (blink1073) [inline src/cells/model.ts]: I'm wondering whether `focused` is the right word at all, maybe `editing`?

Comment (blink1073) [inline src/cells/model.ts]: Because in actuality, the cell itself is focused when not editing.

Comment (dwillmer) [inline src/cells/model.ts]: Sounds sensible

Comment (blink1073) [inline src/cells/model.ts]: Actually, I prefer `mode: 'command' | 'edit'` to correspond with the notebook level attribute.

**Pull request merged**